File: 2.lang/java/basic/src/test/java/org/example/functional/ImmutablePersonTest.java
package org.example.functional;

import org.junit.jupiter.api.Test;

import java.util.ArrayList;

import static org.example.functional.ImmutablePerson.badFunction;
import static org.junit.jupiter.api.Assertions.*;

class ImmutablePersonTest {

    @Test
    void testBadFunction() {
        ImmutablePerson m = new ImmutablePerson(
                "William",
                "rodriguez",
                new ArrayList<>()
        );

        assertThrows(UnsupportedOperationException.class, () -> badFunction(m));
        assertEquals(0, m.emails().size());
    }


    @Test
    void testBadFunction_2() {
        ImmutablePerson_2 m = new ImmutablePerson_2(
                "William",
                "rodriguez",
                new ArrayList<>()
        );

        ImmutablePerson_2.badFunction(m);

        assertEquals(0, m.emails().size());
    }
}

File: 2.lang/kotlin/basics/src/test/kotlin/functional/ImmutablePersonTest.kt
package functional

import kotlin.test.Test
import kotlin.test.assertEquals


class ImmutablePersonTest {
    @Test
    fun testMutablePerson() {
        val person = ImmutablePerson(
            lastName = "rodriguez",
            emails = listOf("mail@mail.com")
        )
        assertEquals(1, person.emails.size)
        assertEquals("rodriguez", person.lastName)
        assertEquals("Work", person.category)
    }
}